### PR TITLE
JSON protocol support

### PR DESCRIPTION
**Pull request by denismakogon**
Closes: #4

### CODE REVIEW

File: examples/json/hello/func.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	var person struct {
+		Name string `json:"name"`
+	}
+	json.NewDecoder(in).Decode(&person)
+
+	if person.Name == "" {
+		person.Name = "world"
+	}
+
+	out.Write([]byte(fmt.Sprintf("Hello %s!\n", person.Name)))
```
Comment (rdallman) [inline examples/json/hello/func.go]: don't we need to return json here?

Comment (rdallman) [reply examples/json/hello/func.go]: i see that the handler does this, so this example is just slightly confusing in that the incoming path decodes a json body but writes back plain text. i understand the protocol underneath is simply an additional layer of json that lets the user write whatever they want in and out, I think this would be slightly less confusing maybe if all plaintext or all json instead of half and half.

File: fdk.go [hunk 2]
```diff
@@ -107,6 +111,84 @@ func doHTTP(handler Handler, ctx context.Context, in io.Reader, out io.Writer) {
...
+		jsonio: jsonio{
+			ContentType: "application/json",
+		},
+		Protocol: &callResponseHTTP{
+			StatusCode: 200,
+		},
+		Writer: buf,
+	}
+
+	req := &jsonIn{}
```
Comment (rdallman) [inline fdk.go]: nit: can save an alloc here by using a var and referencing it (1 alloc each loop)

File: fdk.go [hunk 3]
```diff
@@ -107,6 +111,84 @@ func doHTTP(handler Handler, ctx context.Context, in io.Reader, out io.Writer) {
...
+
+func (out *jsonOut) WriteStatus(status int) {
+	out.Protocol.StatusCode = status
+}
+
+func doJSONOnce(handler Handler, ctx context.Context, in io.Reader, out io.Writer, buf *bytes.Buffer, hdr http.Header) {
+	buf.Reset()
+	resetHeaders(hdr)
+
+	resp := &jsonOut{
```
Comment (rdallman) [inline fdk.go]: nit: ditto on allocs

File: examples/json/hello/func.go (hunk 1 above)
Comment (denismakogon) [reply examples/json/hello/func.go]: no, developers should just write response body using this `.Write`

File: fdk.go (hunk 2 above)
Comment (denismakogon) [reply fdk.go]: makes sense

File: fdk.go [hunk 4]
```diff
@@ -107,6 +111,84 @@ func doHTTP(handler Handler, ctx context.Context, in io.Reader, out io.Writer) {
...
+}
+
+type callResponseHTTP struct {
+	StatusCode int         `json:"status_code,omitempty"`
+	Headers    http.Header `json:"headers,omitempty"`
+}
+
+type jsonOut struct {
+	jsonio
+	Protocol *callResponseHTTP `json:"protocol,omitempty"`
```
Comment (rdallman) [inline fdk.go]: this needn't be a pointer, we'll get better locality if it's a value and from what I can tell this is never nil (the json will always be displayed as an object just the same for a value as a non-nil value pointer)

File: fdk.go [hunk 5]
```diff
@@ -107,6 +111,84 @@ func doHTTP(handler Handler, ctx context.Context, in io.Reader, out io.Writer) {
...
+type callRequestHTTP struct {
+	Type       string      `json:"type"`
+	RequestURL string      `json:"request_url"`
+	Headers    http.Header `json:"headers"`
+}
+
+type jsonIn struct {
+	jsonio
+	CallID   string           `json:"call_id"`
+	Protocol *callRequestHTTP `json:"protocol"`
```
Comment (rdallman) [inline fdk.go]: ditto about pointers, this can easily be checked for `.StatusCode != 0` here for nil-ity

File: fdk.go [hunk 6]
```diff
@@ -107,6 +111,84 @@ func doHTTP(handler Handler, ctx context.Context, in io.Reader, out io.Writer) {
...
+}
+
+type jsonOut struct {
+	jsonio
+	Protocol *callResponseHTTP `json:"protocol,omitempty"`
+
+	io.Writer
+}
+
+func (out *jsonOut) Write(p []byte) (n int, err error) {
```
Comment (rdallman) [inline fdk.go]: the methods around this are a little confusing. one nit is that this [and jsonIn] would be slightly more readable if all the fields were all here instead of inheriting a mix of embedded fields from `jsonio` as well as any methods on `io.Writer`.

this writer seems unused since this struct overwrites `io.Writer` method and does not use the `io.Writer` inside of it. this writer (buffer) seems best used in the caller and not embedded in this struct, to accumulate a body from the user into it (in the caller), which can then be inserted in the `*jsonOut` body as a string once (in the caller), instead of doing many string concatenations (which each generate a new string, strings are immutable). i realize this was handling status code setting as well. personally i like having models that are just models, but maybe that is just me; it would be possible to have another simply type to allow setting status code and writing to the buffer, which would then be translated to a json encodable struct, but having these in the same struct is not the end of the world. at present, however, we need to stop overwriting the Write method and bypassing the buffer, and set the body in doJSONOnce before encoding this struct.

File: examples/json/hello/func.go (hunk 1 above)
Comment (rdallman) [reply examples/json/hello/func.go]: it's just a writer, there is an http example that just does too `json.NewEncoder(out).Encode(&myThing)` -- it just means that even with json protocol, http response to the client will be a json body (many people will want this). at one point we wanted to make all examples JSON. in any event, either plain text for everything or json for everything in the example. the mixed I find really confusing.

File: fdk.go [hunk 7]
```diff
@@ -115,26 +115,25 @@ func doJSON(handler Handler, ctx context.Context, in io.Reader, out io.Writer) {
 	var buf bytes.Buffer
 	hdr := make(http.Header)
 
+	resp := &jsonOut{}
+	req := &jsonIn{}
+
 	for {
-		doJSONOnce(handler, ctx, in, out, &buf, hdr)
+		doJSONOnce(handler, ctx, in, out, resp, req, &buf, hdr)
```
Comment (rdallman) [inline fdk.go]: not sure this is that useful, this function signature is really bloated now. buffer and headers both have sizable maps allocated which are useful to re-use. this is just a pointer. in scope what I meant by using var syntax was:

```go
var resp jsonOut
// ...
doAThing(&resp)
```

Comment (rdallman) [reply fdk.go]: this has bad side effects as well (I can elaborate, if desired)

Comment (denismakogon) [reply fdk.go]: fixed.

File: fdk.go [hunk 8]
```diff
@@ -143,52 +142,41 @@ type callResponseHTTP struct {
...
 	} else {
-		setHeaders(ctx, req.Protocol.Headers)
+		setHeaders(ctx, jsonRequest.Protocol.Headers)
 		//TODO(xxx): use FN_DEADLINE here
 		ctx, cancel := context.WithCancel(ctx)
 		defer cancel()
-		handler.Serve(ctx, strings.NewReader(req.Body), resp)
-		resp.WriteStatus(200)
+		handler.Serve(ctx, strings.NewReader(jsonRequest.Body), buf)
+		jsonResponse.WriteStatus(200)
```
Comment (rdallman) [inline fdk.go]: 2 things with `WriteStatus`:

* our JSON protocol allows a user to send back a custom response code to reply to the client with, here they can only send 200 if their function runs ok (regardless of if they want to return another code) since we aren't checking for the status that the user may have written, to allow that:
* since we're not sending `jsonResponse` to `handler.Serve` the user has no way to set a status code. the way for them to do that is https://github.com/fnproject/fdk-go/blob/master/fdk.go#L63 -- we can easily change this code to:

```go
cw, ok := out.(interface{
  WriteStatus(code int)
})
if ok {
  cw.WriteStatus(code)
}
```

and then this code will need to change to let `jsonResponse` or an object like it that can save a status implement `io.Reader` to return the body as well as allow status writing. open to ideas for better ways to allow `WriteStatus` for the client, what we have seems like we can hide the magic pretty well and make it pretty slick with the outlined methodology.

Comment (denismakogon) [reply fdk.go]: better to reuse response object that is already there, so just needed to adjust json request handler

File: examples/json/hello/func.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+	var person struct {
+		Name string `json:"name"`
+	}
+	json.NewDecoder(in).Decode(&person)
+
+	if person.Name == "" {
+		person.Name = "world"
+	}
+	body := fmt.Sprintf("Hello %s!\n", person.Name)
+	err := json.NewEncoder(out).Encode(body)
```
Comment (rdallman) [inline examples/json/hello/func.go]: it would be nice if this mimicked the other example we have, with an object. this isn't really a json response, it's a string (which will encode, sure)

File: fdk.go [hunk 10]
```diff
@@ -107,6 +111,69 @@ func doHTTP(handler Handler, ctx context.Context, in io.Reader, out io.Writer) {
...
+		jsonResponse.Protocol.StatusCode = 500
+		jsonResponse.Body = fmt.Sprintf(`{"error": %v}`, err.Error())
+	} else {
+		setHeaders(ctx, jsonRequest.Protocol.Headers)
+		//TODO(xxx): use FN_DEADLINE here
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		handler.Serve(ctx, strings.NewReader(jsonRequest.Body), &resp)
+		jsonResponse.Protocol.StatusCode = resp.status
+		jsonResponse.Body = buf.String()
```
Comment (rdallman) [inline fdk.go]: we should set headers from `resp` as well here

Comment (denismakogon) [reply fdk.go]: fixed

File: examples/README.md [hunk 11]
```diff
@@ -1,4 +1,4 @@
-# Hot Function Example
+# Hot JSON Function Example
```
Comment (rdallman) [inline examples/README.md]: it's unclear why this change is here, the example provided gives `--format http`. since that's tested, I think it's fine to leave that as is and undo this edit. since the example will work with http and json since it uses the fdk, i'm not sure it really matters to highlight the specificity anymore

Comment (denismakogon) [reply examples/README.md]: Fixed. Decided to add the link to the fn formats doc which is quite useful in understanding those.
